examples/local/backups/restart_tablets.sh

```diff
@@ -23,19 +23,31 @@ for i in 100 101 102; do
   CELL=zone1 TABLET_UID=$i ./scripts/mysqlctl-up.sh
   CELL=zone1 KEYSPACE=commerce TABLET_UID=$i ./scripts/vttablet-up.sh
 done
-sleep 5
-vtctldclient InitShardPrimary --force commerce/0 zone1-100
 
 for i in 200 201 202; do
   CELL=zone1 TABLET_UID=$i ./scripts/mysqlctl-up.sh
   SHARD=-80 CELL=zone1 KEYSPACE=customer TABLET_UID=$i ./scripts/vttablet-up.sh
 done
-sleep 5
-vtctldclient InitShardPrimary --force customer/-80 zone1-200
 
 for i in 300 301 302; do
   CELL=zone1 TABLET_UID=$i ./scripts/mysqlctl-up.sh
   SHARD=80- CELL=zone1 KEYSPACE=customer TABLET_UID=$i ./scripts/vttablet-up.sh
 done
 sleep 5
+
+# Wait for all the replica tablets to be in the serving state before initiating
+# InitShardPrimary. This is essential, since we want the RESTORE phase to be
+# complete before we start InitShardPrimary, otherwise we end up reading the
+# tablet type to RESTORE and do not set semi-sync, which leads to the primary
+# hanging on writes.
+for i in 101 201 301; do
```

Comment:
the way this is structured, your wait time scales linearly with the number of shards. we could speed this up by waiting ~~ sort of ~~ concurrently:

```bash
declare -a healthy

for _ in $(seq 0 300); do
  for i in 1 2 3; do
    if [[ "${healthy[$i] -eq 1 ]]; then
      continue;
    fi

    status=$(curl "http://$hostname:15${i}01/debug/status_details" | grep "REPLICA: serving")
    if [[ -n "$status" ]]; then
        healthy[$i]=1;
    fi
  done;
  # iterate and check if all 3 are healthy, then break
done;

# fail if not all 3 healthy
```

Reply:
Why would the wait time scale linearly with my original code? The tablets are already started and we are just waiting for them to reach a specific state. The loop for all the tablets after the slowest tablet will just run immediately, right? So we will only be waiting for the time of the slowest tablet.

Reply:
If we want to do this in parallel, we could use gnu parallel. For example, something like this:
```
for _ in $(seq 0 300); do
  cmd=""
  for i in 1 2 3; do
    if [[ "${healthy[$i]}" -eq 1 ]]; then
      continue;
    fi

    cmd+='if curl --max-time 1 "http://$hostname:15${i}01/debug/status_details" | grep -v "REPLICA: serving"; then healthy[${i}]=1; fi; '
  done;
  parallel -nocite ::: eval "${cmd}"
  # iterate and check if all 3 are healthy, then break
done;
```

Reply:
If we want to do this in parallel, we could use gnu parallel. For example, something like this:
```
for _ in $(seq 0 300); do
  cmd=""
  for i in 1 2 3; do
    if [[ "${healthy[$i]}" -eq 1 ]]; then
      continue;
    fi

    cmd+="if curl --max-time 1 \"http://$hostname:15${i}01/debug/status_details\" | grep -q \"REPLICA: serving\"; then healthy[${i}]=1; fi; "
  done;
  echo -n "CMD: ${cmd}"
  parallel -nocite ::: eval "${cmd}"
  # iterate and check if all 3 are healthy, then break
done;
```

Reply:
Note that it's been a while since I used parallel and it's a little esoteric so please don't consider that viable code ^^ :D but just a pointer if we care enough to parallelize it.

Reply:
Thankyou @mattlord for this information. TIL gnu parallel exists!

Reply:
@ajm188 @mattlord I have done something different achieving the same goal, which I find more intuitive to understand. Could you take a look at this again?